<a href="https://colab.research.google.com/github/bbrauser/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Brad_Brauser_LS_DS_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [1]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-06-03 22:15:48--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.28.246
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.28.246|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.2’

instacart_online_gr 100%[===================>] 196.03M   101MB/s    in 2.0s    

2020-06-03 22:15:50 (101 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.2’ saved [205548478/205548478]



In [2]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [3]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [4]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [5]:
##### YOUR CODE HERE #####
import pandas as pd

orders = pd.read_csv('orders.csv')

order_products__prior = pd.read_csv('order_products__prior.csv')

products = pd.read_csv('products.csv')

top_df = [orders, order_products__prior, products]

pd.concat(top_df)

order_id  user_id  ... aisle_id  department_id
0      2539329.0      1.0  ...      NaN            NaN
1      2398795.0      1.0  ...      NaN            NaN
2       473747.0      1.0  ...      NaN            NaN
3      2254736.0      1.0  ...      NaN            NaN
4       431534.0      1.0  ...      NaN            NaN
...          ...      ...  ...      ...            ...
49683        NaN      NaN  ...    124.0            5.0
49684        NaN      NaN  ...     42.0            1.0
49685        NaN      NaN  ...    112.0            3.0
49686        NaN      NaN  ...     41.0            8.0
49687        NaN      NaN  ...     73.0           11.0

[35905260 rows x 13 columns]

## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [6]:
'We can join/concatante two strings together ' + 'using the "+" operator.'

'We can join/concatante two strings together using the "+" operator.'

In [0]:
import pandas as pd

In [0]:
df1 = pd.DataFrame({'a': [1,2,3,4], 'b': [4,5,6,7], 'c': [7,8,9,10]})

df2 = pd.DataFrame({'a': [6,4,8,7], 'b': [9,4,3,2], 'c': [1,6,2,9]})

In [9]:
df1.head()

a  b   c
0  1  4   7
1  2  5   8
2  3  6   9
3  4  7  10

In [10]:
df2.head()

a  b  c
0  6  9  1
1  4  4  6
2  8  3  2
3  7  2  9

In [11]:
by_rows = pd.concat([df1, df2])

by_rows

a  b   c
0  1  4   7
1  2  5   8
2  3  6   9
3  4  7  10
0  6  9   1
1  4  4   6
2  8  3   2
3  7  2   9

In [12]:
by_rows.reset_index()

index  a  b   c
0      0  1  4   7
1      1  2  5   8
2      2  3  6   9
3      3  4  7  10
4      0  6  9   1
5      1  4  4   6
6      2  8  3   2
7      3  7  2   9

In [13]:
by_cols = pd.concat([df1,df2], axis=1)

by_cols

a  b   c  a  b  c
0  1  4   7  6  9  1
1  2  5   8  4  4  6
2  3  6   9  8  3  2
3  4  7  10  7  2  9

In [14]:
by_cols['a']

a  a
0  1  6
1  2  4
2  3  8
3  4  7

In [0]:
by_cols.columns = ['a1', 'b1', 'c1', 'a2', 'b2', 'c2']

In [16]:
by_cols['a1']

0    1
1    2
2    3
3    4
Name: a1, dtype: int64

In [0]:
df3 = pd.DataFrame({'a': [4,3,2,1], 'b': [4,5,6,7], 'c': [7,8,9,10]})

df4 = pd.DataFrame({'a': [6,4,8,7,8], 'b': [9,4,3,2,1], 'd': [1,6,2,9,5]})

In [18]:
df3.head()

a  b   c
0  4  4   7
1  3  5   8
2  2  6   9
3  1  7  10

In [19]:
df4.head()

a  b  d
0  6  9  1
1  4  4  6
2  8  3  2
3  7  2  9
4  8  1  5

In [20]:
by_rows = pd.concat([df3,df4])

by_rows

a  b     c    d
0  4  4   7.0  NaN
1  3  5   8.0  NaN
2  2  6   9.0  NaN
3  1  7  10.0  NaN
0  6  9   NaN  1.0
1  4  4   NaN  6.0
2  8  3   NaN  2.0
3  7  2   NaN  9.0
4  8  1   NaN  5.0

In [21]:
by_cols = pd.concat([df3,df4], axis = 'columns')

by_cols

a    b     c  a  b  d
0  4.0  4.0   7.0  6  9  1
1  3.0  5.0   8.0  4  4  6
2  2.0  6.0   9.0  8  3  2
3  1.0  7.0  10.0  7  2  9
4  NaN  NaN   NaN  8  1  5

In [22]:
from IPython.display import display, Image
url = 'https://cdn-images-1.medium.com/max/1600/1*vYGFQCafJtGBBX5mbl0xyw.png'
example = Image(url=url, width=500)

display(example)

In [23]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-06-03 22:16:17--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.107.222
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.107.222|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’

instacart_online_gr 100%[===================>] 196.03M  96.3MB/s    in 2.0s    

2020-06-03 22:16:19 (96.3 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’ saved [205548478/205548478]



In [24]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [25]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01/instacart_2017_05_01


In [26]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


In [27]:
display(example)

In [28]:
aisles = pd.read_csv('aisles.csv')
print(aisles.shape)
aisles.head()

(134, 2)


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [29]:
departments = pd.read_csv('departments.csv')
print(departments.shape)
departments.head()

(21, 2)


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [30]:
order_products__prior = pd.read_csv('order_products__prior.csv')
print(order_products__prior.shape)
order_products__prior.head()

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [31]:
order_products__train = pd.read_csv('order_products__train.csv')
print(order_products__train.shape)
order_products__train.head()

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [32]:
orders = pd.read_csv('orders.csv')
print(orders.shape)
orders.head()

(3421083, 7)


order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [33]:
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [34]:
order_products = pd.concat([order_products__prior, order_products__train])

print(order_products.shape)
order_products.head()

(33819106, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [35]:
order_products__prior.shape

(32434489, 4)

In [36]:
order_products__train.shape

(1384617, 4)

In [37]:
order_products.shape

(33819106, 4)

In [38]:
display(example)

In [39]:
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [40]:
order_products.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [41]:
products.head()

product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [42]:
display(example)

In [43]:
condition = orders['order_id'] <= 5
orders[condition].head(5)

order_id  user_id  ... order_hour_of_day  days_since_prior_order
1868044         1   112108  ...                10                     9.0
2593373         5   156122  ...                16                     9.0
2958007         4   178520  ...                 9                     7.0
3355525         2   202279  ...                 9                     8.0
3417191         3   205970  ...                17                    12.0

[5 rows x 7 columns]

In [44]:
orders[orders['order_id'] <= 5].head(5)

order_id  user_id  ... order_hour_of_day  days_since_prior_order
1868044         1   112108  ...                10                     9.0
2593373         5   156122  ...                16                     9.0
2958007         4   178520  ...                 9                     7.0
3355525         2   202279  ...                 9                     8.0
3417191         3   205970  ...                17                    12.0

[5 rows x 7 columns]

In [45]:
order_products['order_id'] == 2539329

0          False
1          False
2          False
3          False
4          False
           ...  
1384612    False
1384613    False
1384614    False
1384615    False
1384616    False
Name: order_id, Length: 33819106, dtype: bool

In [46]:
order_products_subset = order_products[(order_products['order_id'] == 2539329) | (order_products['order_id'] == 2398795)]

order_products_subset.head(11)

order_id  product_id  add_to_cart_order  reordered
22742744   2398795         196                  1          1
22742745   2398795       10258                  2          0
22742746   2398795       12427                  3          1
22742747   2398795       13176                  4          0
22742748   2398795       26088                  5          1
22742749   2398795       13032                  6          0
24076664   2539329         196                  1          0
24076665   2539329       14084                  2          0
24076666   2539329       12427                  3          0
24076667   2539329       26088                  4          0
24076668   2539329       26405                  5          0

In [47]:
orders_subset = orders[(orders['user_id'] == 1) & (orders['order_number'] <= 2)]

order_products_subset.head(11)

order_id  product_id  add_to_cart_order  reordered
22742744   2398795         196                  1          1
22742745   2398795       10258                  2          0
22742746   2398795       12427                  3          1
22742747   2398795       13176                  4          0
22742748   2398795       26088                  5          1
22742749   2398795       13032                  6          0
24076664   2539329         196                  1          0
24076665   2539329       14084                  2          0
24076666   2539329       12427                  3          0
24076667   2539329       26088                  4          0
24076668   2539329       26405                  5          0

In [48]:
orders_and_products = pd.merge(orders_subset, order_products_subset, on='order_id', how='inner')

orders_and_products.head(11)

order_id  user_id eval_set  ...  product_id  add_to_cart_order  reordered
0    2539329        1    prior  ...         196                  1          0
1    2539329        1    prior  ...       14084                  2          0
2    2539329        1    prior  ...       12427                  3          0
3    2539329        1    prior  ...       26088                  4          0
4    2539329        1    prior  ...       26405                  5          0
5    2398795        1    prior  ...         196                  1          1
6    2398795        1    prior  ...       10258                  2          0
7    2398795        1    prior  ...       12427                  3          1
8    2398795        1    prior  ...       13176                  4          0
9    2398795        1    prior  ...       26088                  5          1
10   2398795        1    prior  ...       13032                  6          0

[11 rows x 10 columns]

In [49]:
condition = products['product_id'].isin(orders_and_products['product_id'])

products_subset = products[condition]

products_subset.head(10)

product_id  ... department_id
195           196  ...             7
10257       10258  ...            19
12426       12427  ...            19
13031       13032  ...            14
13175       13176  ...             4
14083       14084  ...            16
26087       26088  ...            19
26404       26405  ...            17

[8 rows x 4 columns]

In [50]:
products_subset = products_subset[['product_id', 'product_name']]

products_subset.head()

product_id            product_name
195           196                    Soda
10257       10258              Pistachios
12426       12427     Original Beef Jerky
13031       13032   Cinnamon Toast Crunch
13175       13176  Bag of Organic Bananas

In [51]:
final = pd.merge(products_subset, orders_and_products, on='product_id', how='inner')

final.head(11)

product_id  ... reordered
0          196  ...         0
1          196  ...         1
2        10258  ...         0
3        12427  ...         0
4        12427  ...         1
5        13032  ...         0
6        13176  ...         0
7        14084  ...         0
8        26088  ...         0
9        26088  ...         1
10       26405  ...         0

[11 rows x 11 columns]

In [52]:
display(example)

In [53]:
final = final.sort_values(by=['order_number', 'add_to_cart_order'])

final.head(11)

product_id  ... reordered
0          196  ...         0
7        14084  ...         0
3        12427  ...         0
8        26088  ...         0
10       26405  ...         0
1          196  ...         1
2        10258  ...         0
4        12427  ...         1
6        13176  ...         0
9        26088  ...         1
5        13032  ...         0

[11 rows x 11 columns]

In [54]:
final = final[['user_id', 'order_id', 'order_number','order_dow', 'order_hour_of_day', 'add_to_cart_order', 'product_id', 'product_name']]

final.head(11)

user_id  order_id  ...  product_id                             product_name
0         1   2539329  ...         196                                     Soda
7         1   2539329  ...       14084  Organic Unsweetened Vanilla Almond Milk
3         1   2539329  ...       12427                      Original Beef Jerky
8         1   2539329  ...       26088               Aged White Cheddar Popcorn
10        1   2539329  ...       26405         XL Pick-A-Size Paper Towel Rolls
1         1   2398795  ...         196                                     Soda
2         1   2398795  ...       10258                               Pistachios
4         1   2398795  ...       12427                      Original Beef Jerky
6         1   2398795  ...       13176                   Bag of Organic Bananas
9         1   2398795  ...       26088               Aged White Cheddar Popcorn
5         1   2398795  ...       13032                    Cinnamon Toast Crunch

[11 rows x 8 columns]

In [55]:
final.columns = [column.replace('_', ' ') for column in final]

final.head(11)

user id  order id  ...  product id                             product name
0         1   2539329  ...         196                                     Soda
7         1   2539329  ...       14084  Organic Unsweetened Vanilla Almond Milk
3         1   2539329  ...       12427                      Original Beef Jerky
8         1   2539329  ...       26088               Aged White Cheddar Popcorn
10        1   2539329  ...       26405         XL Pick-A-Size Paper Towel Rolls
1         1   2398795  ...         196                                     Soda
2         1   2398795  ...       10258                               Pistachios
4         1   2398795  ...       12427                      Original Beef Jerky
6         1   2398795  ...       13176                   Bag of Organic Bananas
9         1   2398795  ...       26088               Aged White Cheddar Popcorn
5         1   2398795  ...       13032                    Cinnamon Toast Crunch

[11 rows x 8 columns]

In [56]:
display(example)

In [84]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2.reset_index()

index  John Smith  Jane Doe  Mary Johnson
0  treatmenta         NaN      16.0           3.0
1  treatmentb         2.0      11.0           1.0

In [85]:
# table2 -> tidy

tidy = table2.melt(id_vars='index', value_vars=['treatmenta', 'treatmentb'])

tidy1 = tidy.rename(columns={
    'index': 'name', 
    'variable': 'trt', 
    'value': 'result'
})

tidy1

KeyError: ignored

In [77]:
# tidy -> table2

wide = tidy.pivot_table(index='name', columns='trt', values='result')

wide

KeyError: ignored

In [0]:
flights = sns.load_dataset('flights')

In [0]:
# Flights Pivot Table

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####